src/Core/src/Signer/SignerV4.php

```diff
@@ -56,87 +57,202 @@ public function __construct(string $scopeName, string $region)
         $this->region = $region;
     }
 
-    public function sign(Request $request, ?Credentials $credentials): void
+    public function presign(Request $request, Credentials $credentials, RequestContext $context): void
     {
-        if (null === $credentials) {
-            return;
+        $now = $context->getCurrentDate() ?? new \DateTimeImmutable();
+        $expires = $context->getExpirationDate() ?? (new \DateTimeImmutable($now->format(\DateTimeInterface::ATOM)))->add(new \DateInterval('PT1H'));
+
+        $this->handleSignature($request, $credentials, $now, $expires, true);
+    }
+
+    public function sign(Request $request, Credentials $credentials, RequestContext $context): void
+    {
+        $now = $context->getCurrentDate() ?? new \DateTimeImmutable();
+
+        $this->handleSignature($request, $credentials, $now, $now, false);
+    }
+
+    protected function buildBodyDigest(Request $request, bool $isPresign): string
+    {
+        if ($request->hasHeader('x-amz-content-sha256')) {
+            $hash = ((array) $request->getHeader('x-amz-content-sha256'))[0];
+        } else {
+            $hash = $request->getBody()->hash();
+        }
+
+        if ('UNSIGNED-PAYLOAD' === $hash) {
+            $request->setHeader('x-amz-content-sha256', $hash);
+        }
+
+        return $hash;
+    }
+
+    private function handleSignature(Request $request, Credentials $credentials, \DateTimeInterface $now, \DateTimeInterface $expires, bool $isPresign): void
+    {
+        $this->removePresign($request);
+        $this->sanitizeHostForHeader($request);
+        $this->assignAmzQueryValues($request, $credentials, $isPresign);
+
+        $this->buildTime($request, $now, $expires, $isPresign);
+        $credentialScope = $this->buildCredentialString($request, $credentials, $now, $isPresign);
+        $credentialString = \implode('/', $credentialScope);
+        $signingKey = $this->buildSigningKey($credentials, $credentialScope);
+
+        // signature is passed by reference to convertBodyToStream
+        $signature = '';
```

Comment:
This variable is never used elsewhere than in the `convertBodyToStream` call (it is then always overridden on line 120 before being used again). So why defining it outside the `if/else` block ?

Reply:
indeed, the $signature could have been initialized in the `else` branch.
It have to be defined before the call to `convertBodyToStream` for matching the typehinting.

Reply:
Shouldn't we actually move the line 120 here, to pass the computed signature instead of computing it twice ?

Reply:
sadly not, there are complex circular dependency.
The line 120 `buildSignature` need the headers and the headers may be updated by this `buildBodyDigest`.